### PR TITLE
Fix error handling in searchLoclImages()

### PR DESCRIPTION
**Pull request by prologic**
Hopefully this improves our ability to debug the masked error and potentially racey behaviour in infra that relies on the Docker provider via Terraform mover SSH.

cc @Junkern

### CODE REVIEW

File: internal/provider/resource_docker_network.go [hunk 1]
```diff
@@ -155,51 +151,3 @@ func resourceDockerNetwork() *schema.Resource {
 		},
 	}
 }
-
```
Comment (Junkern) [inline internal/provider/resource_docker_network.go:158]: Could you re-add the deleted functions or is there a reason you deleted them?

File: internal/provider/resource_docker_network_test.go [hunk 2]
```diff
@@ -325,48 +325,6 @@ func testAccNetworkIPv4(network *types.NetworkResource, internal bool) resource.
 	}
 }
 
-func TestAccDockerNetwork_ipv6(t *testing.T) {
```
Comment (Junkern) [inline internal/provider/resource_docker_network_test.go:328]: Same with the tests: Is there a reason you deleted them?

File: internal/provider/resource_docker_service_test.go [hunk 3]
```diff
@@ -437,7 +437,7 @@ func TestAccDockerService_fullSpec(t *testing.T) {
 			s.Spec.TaskTemplate.ContainerSpec.Secrets[0].File.UID != "0" ||
 			s.Spec.TaskTemplate.ContainerSpec.Secrets[0].File.GID != "0" ||
 			// nolint: staticcheck
-			s.Spec.TaskTemplate.ContainerSpec.Secrets[0].File.Mode != os.FileMode(777) {
```
Comment (Junkern) [inline internal/provider/resource_docker_service_test.go:440]: I quickly looked into the code, and it seems like the whole parsing of the FileMode option is weird. I would ask you to leave this out of this PR and create an issue for that, so we can tackle this at a later point

File: internal/provider/resource_docker_image_funcs.go [hunk 4]
```diff
@@ -106,20 +110,23 @@ func resourceDockerImageDelete(ctx context.Context, d *schema.ResourceData, meta
 }
 
 // Helpers
-func searchLocalImages(ctx context.Context, client *client.Client, data Data, imageName string) *types.ImageSummary {
+func searchLocalImages(ctx context.Context, client *client.Client, data Data, imageName string) (*types.ImageSummary, error) {
 	imageInspect, _, err := client.ImageInspectWithRaw(ctx, imageName)
 	if err != nil {
-		return nil
+		return nil, fmt.Errorf("unable to inspect imave %s: %w", imageName, err)
```
Comment (Junkern) [inline internal/provider/resource_docker_image_funcs.go:116]:
```suggestion
		return nil, fmt.Errorf("unable to inspect image %s: %w", imageName, err)
```

File: internal/provider/resource_docker_service_structures.go [hunk 5]
```diff
@@ -1083,17 +1083,14 @@ func createServiceNetworks(v interface{}) ([]swarm.NetworkAttachmentConfig, erro
 func createLogDriver(v interface{}) (*swarm.Driver, error) {
 	logDriver := swarm.Driver{}
 	if len(v.([]interface{})) > 0 {
-		for _, rawLogging := range v.([]interface{}) {
```
Comment (Junkern) [inline internal/provider/resource_docker_service_structures.go:1086]: Why did you remove the loop?

**Review by Junkern — CHANGES_REQUESTED** [review]

File: internal/provider/resource_docker_network.go (hunk 1 above)
Comment (prologic) [reply internal/provider/resource_docker_network.go:158]: Ahh yes sorry I can. I deleted it because of my own OCD when the linter yells at my for "unused" functions 😅

Comment (prologic) [reply internal/provider/resource_docker_network.go]: Done!

File: internal/provider/resource_docker_service_test.go (hunk 3 above)
Comment (prologic) [reply internal/provider/resource_docker_service_test.go:440]: Okay! One sec...

Comment (prologic) [reply internal/provider/resource_docker_service_test.go]: Fix removed from this patch

File: internal/provider/resource_docker_service_structures.go (hunk 5 above)
Comment (prologic) [reply internal/provider/resource_docker_service_structures.go:1086]: If you have a look at the TODO(s) there and the logic, I _think_ it only goes thorugh the loop once anyway, so the loop is superfluous. Linter also says so 😅

**Review by Junkern — CHANGES_REQUESTED** [review]
One small change. Afterwards we can merge (if all test pass)

File: internal/provider/resource_docker_service_structures.go (hunk 5 above)
Comment (prologic) [reply internal/provider/resource_docker_service_structures.go:1086]: Can you link me to where you found this `MAX_ELEMENTS: 1`? I'm getting a bit lost here in the code, api docs and what not 😅

Comment (Junkern) [reply internal/provider/resource_docker_service_structures.go:1086]: https://github.com/kreuzwerker/terraform-provider-docker/blob/master/internal/provider/resource_docker_service.go#L687

**Comment by prologic** [comment]
> One small change. Afterwards we can merge (if all test pass). I have the feeling by changing the error handling we might have to adapt a few tests as well, but lets see. (You can also run them locally)

I _might_ need your help here,  I'm not actually sure why the tests pass for some versions of Go but fails for others.

I've ran `TestAccDockerService_convergeAndStopGracefully` locally here on my workstation and it passes just fine. Hmmm ? 🤔

I _hope_ I haven't changed the behaviour of anything!

**Comment by Junkern** [comment]
We did indeed change the behaviour. Before, if the `ImageInspectWithRaw` call within the `searchLocalImages` did not find any image, it simply did not error, and the parentFunction would then continue searching for the image in the registry.

Now the implementation fails as soon as an image cannot be found locally. But sometimes the image is not stored locally. I tested the following implementation and it seems to work for a few selected tests (I cannot push to your branch):
```go
func searchLocalImages(ctx context.Context, client *client.Client, data Data, imageName string) (*types.ImageSummary, error) {
	imageInspect, _, err := client.ImageInspectWithRaw(ctx, imageName)
	if err != nil {
	        // the if below is new
		if errdefs.IsNotFound(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("unable to inspect image %s: %w", imageName, err)
	}
```

And you also have to `go get github.com/docker/docker/errdefs` to include the correct package

**Comment by Junkern** [comment]
I did include the `searchLocalImage` changes in the newest `v2.18.1` version (which I really wanted to release today).
The changes regarding the `internal/provider/resource_docker_service_structures.go` file will be part of one of the next releases :)

Thanks for contributing! I hope the better error handling helps you

**Comment by prologic** [comment]
@Junkern Thank you so much! Sorry I didn't quite get this over the line, yesterday for me was. nuts, over 6hrs of meetings, didn't get any "real work" done 😅
